storage/cloudspanner/log_storage.go

```diff
@@ -753,12 +810,43 @@ WHERE sd.TreeID = @tree_id AND sd.SequenceNumber >= @start AND sd.SequenceNumber
 		count = xend - start
 	}
 	stmt.Params["xend"] = xend
+	seqLeaves := make(map[string]sequencedLeafDataCols)
+	if err := tx.stx.Query(ctx, stmt).Do(func(r *spanner.Row) error {
+		var seqLeaf sequencedLeafDataCols
+		if err := r.ToStruct(&seqLeaf); err != nil {
+			return err
+		}
+		seqLeaves[string(seqLeaf.LeafIdentityHash)] = seqLeaf
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	idHashes := make([][]byte, 0, len(seqLeaves))
+	for _, l := range seqLeaves {
+		idHashes = append(idHashes, l.LeafIdentityHash)
+	}
 
-	// Results need to be returned in order [start, end), all of which should be
-	// available (as we restricted xend/count to TreeSize).
+	stmt = spanner.NewStatement(
+		`SELECT 
+		   TreeID,
+		   LeafIdentityHash, 
+		   LeafValue, 
+		   ExtraData, 
+		   QueueTimestampNanos
+		 FROM 
+		   LeafData
+		 WHERE 
+		   TreeID = @tree_id AND 
+		   LeafIdentityHash IN UNNEST(@id_hashes)`)
+	stmt.Params["tree_id"] = tx.treeID
+	stmt.Params["id_hashes"] = idHashes
+
+	// Results need to be returned in order [start, end), all of which
+	// should be available (as we restricted xend/count to TreeSize).
 	leaves := make(leafmap)
-	rows := tx.stx.Query(ctx, stmt)
-	if err := rows.Do(leaves.addFullRow); err != nil {
+	if err := tx.stx.Query(ctx, stmt).
+		Do(leaves.addFullRow(seqLeaves)); err != nil {
 		return nil, err
 	}
```

Comment:
This seems identical to lines 725-745. Maybe factor out a function?

Also, lines 813-823 are repeated as well, so could be included in that function.